src/query/const_score_query.rs

```diff
@@ -0,0 +1,178 @@
+use std::collections::BTreeMap;
+use std::fmt;
+
+use crate::query::{Explanation, Query, Scorer, Weight};
+use crate::{DocId, DocSet, Score, Searcher, SegmentReader, TantivyError, Term};
+
+/// `ConstScoreQuery` is a wrapper over a query to provide a constant score.
+/// It can avoid unnecessary score computation on the wrapped query.
+///
+/// The document set matched by the `ConstScoreQuery` is strictly the same as the underlying query.
+/// The configured score is used for each document.
+pub struct ConstScoreQuery {
+    query: Box<dyn Query>,
+    score: Score,
+}
+
+impl ConstScoreQuery {
+    /// Builds a const score query.
+    pub fn new(query: Box<dyn Query>, score: Score) -> ConstScoreQuery {
+        ConstScoreQuery { query, score }
+    }
+}
+
+impl Clone for ConstScoreQuery {
+    fn clone(&self) -> Self {
+        ConstScoreQuery {
+            query: self.query.box_clone(),
+            score: self.score,
+        }
+    }
+}
+
+impl fmt::Debug for ConstScoreQuery {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Const(score={}, query={:?})", self.score, self.query)
+    }
+}
+
+impl Query for ConstScoreQuery {
+    fn weight(&self, searcher: &Searcher, scoring_enabled: bool) -> crate::Result<Box<dyn Weight>> {
+        let inner_weight = self.query.weight(searcher, scoring_enabled)?;
+        Ok(if scoring_enabled {
+            Box::new(ConstWeight::new(inner_weight, self.score))
+        } else {
+            inner_weight
+        })
+    }
+
+    fn query_terms(&self, terms: &mut BTreeMap<Term, bool>) {
+        self.query.query_terms(terms)
+    }
+}
+
+struct ConstWeight {
+    weight: Box<dyn Weight>,
+    score: Score,
+}
+
+impl ConstWeight {
+    pub fn new(weight: Box<dyn Weight>, score: Score) -> Self {
+        ConstWeight { weight, score }
+    }
+}
+
+impl Weight for ConstWeight {
+    fn scorer(&self, reader: &SegmentReader, boost: Score) -> crate::Result<Box<dyn Scorer>> {
+        let inner_scorer = self.weight.scorer(reader, boost)?;
+        Ok(Box::new(ConstScorer::new(inner_scorer, boost * self.score)))
+    }
+
+    fn explain(&self, reader: &SegmentReader, doc: u32) -> crate::Result<Explanation> {
+        let mut scorer = self.scorer(reader, 1.0)?;
+        if scorer.seek(doc) != doc {
+            return Err(TantivyError::InvalidArgument(format!(
+                "Document #({}) does not match",
+                doc
+            )));
+        }
+        let mut explanation =
```

Comment:
Should this just be

```rust
let mut explanation = Explanation::new("Const", self.score);
```

Reply:
Either way works for me.

Reply:
Removing the inner score from here, it gets included in detail anyway